### PR TITLE
DOCS-339: MinIO ILM Migration + Tiering

### PR DESCRIPTION
**Pull request by ravindk89**
# Summary

Create and update reference documentation + tutorials for MinIO ILM + Remote Tiers

# Goals
- Refresh `mc ilm` for missed updates
- Add `mc admin remote` command
- Add tutorials for transitioning objects to s3, gcs, azure
- Add tutorial for object expiration
- Create landing page for object lifecycle management

# Non-Goals

- Complete cleanup of reference pages

# Build Instructions

The MinIO docs require Python 3.9+ and Node for building. For Node, use v14.5.0 or later.

1) Clone the repo, Checkout the branch (`git checkout -B installation-fixes origin/DOCS-339`)
2) Run `npm install` from the branch
3) Create a python venv (`python -m venv venv`). You can skip this step if it already exists
4) Activate the venv (`source venv/bin/activate`)
5) Install requirements (`pip install -r requirements.txt`)
6) Run `make html`
7) Run `python -m http.server --directory build/DOCS-339/html`
8) Open 127.0.0.1:8000 in your browser and take a poke around.

If you run into an error running `python -m http.server` related to `--directory`, you're python version is likely older. CD into the `build/<branch>/html` directory and run the command without the `--directory` argument.

Errors related to `wheel` are likely due to missing `wheel` + other OS-specific dependencies. See https://stackoverflow.com/a/59596814

# Key URLS

- http://127.0.0.1:8000/lifecycle-management/lifecycle-management-overview.html
- http://127.0.0.1:8000/lifecycle-management/create-lifecycle-management-expiration-rule.html
- http://127.0.0.1:8000/lifecycle-management/transition-objects-to-s3.html
- http://127.0.0.1:8000/lifecycle-management/transition-objects-to-gcs.html
- http://127.0.0.1:8000/lifecycle-management/transition-objects-to-azure.html
- http://127.0.0.1:8000/reference/minio-cli/minio-mc/mc-ilm.html#mc-ilm-add
- http://127.0.0.1:8000/reference/minio-cli/minio-mc-admin/mc-admin-tier.html

### CODE REVIEW

File: source/includes/common-minio-tiering.rst [hunk 1]
```diff
@@ -0,0 +1,97 @@
+.. start-create-transition-rule-desc
+
+Use the :mc-cmd:`mc ilm add` command to create a new transition rule
```
Comment (ravindk89) [inline source/includes/common-minio-tiering.rst:3]: This chunk gets included in all three of the transition tutorials as part of the step for creating the new transition rules.

File: source/includes/common-minio-tiering.rst [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+       ignore noncurrent object versions.
+
+       This option has no effect on non-versioned buckets.
+
+     
+.. end-create-transition-rule-desc
+
+.. start-create-transition-user-desc
+
+This step creates users and policies on the MinIO deployment for supporting
```
Comment (ravindk89) [inline source/includes/common-minio-tiering.rst:73]: This chunk gets included in the step where users create users/policies.

File: source/includes/common-minio-tiering.rst [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+The following example uses ``Alpha`` as a placeholder :mc:`alias <mc alias>` for
+the MinIO deployment. Replace this value with the appropriate alias for the
+MinIO deployment on which you are configuring lifecycle management rules.
+Replace the password ``LongRandomSecretKey`` with a long, random, and secure
+secret key as per your organizations best practices for password generation.
+
+.. code-block:: shell
+   :class: copyable
+
+   wget -O - https://docs.min.io/minio/baremetal/examples/LifecycleManagementAdmin.json | \
```
Comment (ravindk89) [inline source/includes/common-minio-tiering.rst:87]: This wont work until I merge to master :/ sorry!

Comment (ravindk89) [reply source/includes/common-minio-tiering.rst:87]: If you want to cheat, change the URL to the python server root instead

**Review by poornas — CHANGES_REQUESTED** [review]

File: source/lifecycle-management/lifecycle-management-overview.rst [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+.. default-domain:: minio
+
+.. contents:: Table of Contents
+   :local:
+   :depth: 1
+
+MinIO Object Lifecycle Management allows creating rules for time or date
+based automatic transition or expiry of objects. For object transition,
+MinIO automatically moves the object to a configured remote S3 storage
```
Comment (krisis) [inline source/lifecycle-management/lifecycle-management-overview.rst:15]:
```suggestion
MinIO automatically moves the object to a configured remote storage
```

File: source/lifecycle-management/lifecycle-management-overview.rst [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+tier. For object expiry, MinIO automatically deletes the object.
+
+MinIO lifecycle management is built for behavior and syntax compatibility with 
+:s3-docs:`AWS S3 Lifecycle Management <object-lifecycle-mgmt.html>`. For
+example, you can export S3 lifecycle management rules and import them into
+MinIO or vice-versa. MinIO uses JSON to describe lifecycle management rules,
+and conversion to or from XMl may be required.
```
Comment (krisis) [inline source/lifecycle-management/lifecycle-management-overview.rst:22]: nit: typo
```suggestion
and conversion to or from XML may be required. 
```

File: source/lifecycle-management/lifecycle-management-overview.rst [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+:mc-cmd-option:`~mc ilm add noncurrentversion-transition-days` option
+when creating the transition rule. 
+
+.. _minio-lifecycle-management-expiration:
+
+Object Expiration
+-----------------
+
+MinIO lifecycle management supports expiring objects on a bucket. Object
+"expiration" involves performing a ``DELETE`` operation on the bucket. For
```
Comment (krisis) [inline source/lifecycle-management/lifecycle-management-overview.rst:80]:
```suggestion
"expiration" involves performing a ``DELETE`` operation on the object. For 
```

File: source/lifecycle-management/transition-objects-to-azure.rst [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+
+Required Azure Permissions
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Object transition lifecycle management rules require additional permissions on
+the remote storage tier. Specifically, MinIO requires the 
+:abbr:`Azure (Microsoft Azure)` credentials provide read, write, list, and
+delete permissions for the remote bucket.
+
+Defer to the `Azure RBAC
```
Comment (krisis) [inline source/lifecycle-management/transition-objects-to-azure.rst:73]: nit: typo
```suggestion
Refer to the `Azure RBAC
```

File: source/lifecycle-management/lifecycle-management-overview.rst [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+MinIO manages retrieving tiered objects on-the-fly without any additional
+application-side logic. MinIO retains the minimum object metadata required to
+support retrieving objects transitioned to a remote tier. MinIO therefore
+*requires* exclusive access to the data on the remote storage tier. Object
+retrieval assumes no external mutation, migration, or deletion of stored
+objects. Object transition is *not* a replacement for backup/recovery 
+strategies such as :ref:`minio-bucket-replication`.
+
+Use the :mc-cmd:`mc admin tier` command to create a remote target for tiering
+data to a supported Cloud Service Provider storage backend. You can then use the
```
Comment (krisis) [inline source/lifecycle-management/lifecycle-management-overview.rst:52]: Suggestion
```suggestion
data to a supported Cloud Service Provider object storage You can then use the
```

File: source/reference/minio-cli/minio-mc-admin/mc-admin-tier.rst [hunk 9]
```diff
@@ -0,0 +1,387 @@
...
+      *must* resolve to the provider specified to
+      :mc-cmd:`~mc admin tier add TIER_TYPE`. 
+
+   .. mc-cmd:: access-key
+      :option:
+
+      *Required*
+      
+      The access key for a user on the remote S3 tier. The user must
+      have permission to perform read/write/list operations on the
```
Comment (krisis) [inline source/reference/minio-cli/minio-mc-admin/mc-admin-tier.rst:181]:
```suggestion
      have permission to perform read/write/list/delete operations on the 
```

File: source/reference/minio-cli/minio-mc-admin/mc-admin-tier.rst [hunk 10]
```diff
@@ -0,0 +1,387 @@
...
+      Required if :mc-cmd:`~mc admin tier add TIER_TYPE` is ``s3``. 
+      This option has no effect for any other value of ``TIER_TYPE``.
+
+   .. mc-cmd:: account-name
+      :option:
+
+      *Required*
+      
+      The account name for a user on the remote Azure tier. The user
+      must have permission to perform read/write/list operations on the
```
Comment (krisis) [inline source/reference/minio-cli/minio-mc-admin/mc-admin-tier.rst:203]:
```suggestion
      must have permission to perform read/write/list/delete operations on the
```

File: source/reference/minio-cli/minio-mc-admin/mc-admin-tier.rst [hunk 11]
```diff
@@ -0,0 +1,387 @@
...
+
+   .. mc-cmd:: credentials-file
+      :option:
+
+      *Required*
+      
+      The `credential file
+      <https://cloud.google.com/docs/authentication/getting-started>`__ for a
+      user on the remote GCS tier. The user must have permission to perform
+      read/write/list operations on the remote bucket or bucket prefix.
```
Comment (krisis) [inline source/reference/minio-cli/minio-mc-admin/mc-admin-tier.rst:227]:
```suggestion
      read/write/list/delete operations on the remote bucket or bucket prefix.
```

File: source/reference/minio-cli/minio-mc-admin/mc-admin-tier.rst [hunk 12]
```diff
@@ -0,0 +1,387 @@
...
+      corresponds to the :mc-cmd:`mc admin tier add TIER_NAME` specified
+      when creating the remote tier.
+
+   .. mc-cmd:: access-key
+      :option:
+
+      *Optional*
+      
+      The access key for a user on the remote S3 tier. The user must
+      have permission to perform read/write/list operations on the
```
Comment (krisis) [inline source/reference/minio-cli/minio-mc-admin/mc-admin-tier.rst:312]:
```suggestion
      have permission to perform read/write/list/delete operations on the 
```

File: source/reference/minio-cli/minio-mc-admin/mc-admin-tier.rst [hunk 13]
```diff
@@ -0,0 +1,387 @@
...
+      :mc-cmd:`~mc admin tier add TIER_TYPE` is ``s3``. 
+      This option has no effect for any other ``TIER_TYPE``.
+
+   .. mc-cmd:: account-
+      :option:
+
+      *Required*
+      
+      The account name for a user on the remote Azure tier. The user
+      must have permission to perform read/write/list operations on the
```
Comment (krisis) [inline source/reference/minio-cli/minio-mc-admin/mc-admin-tier.rst:336]:
```suggestion
      must have permission to perform read/write/list/delete operations on the
```

File: source/reference/minio-cli/minio-mc-admin/mc-admin-tier.rst [hunk 14]
```diff
@@ -0,0 +1,387 @@
...
+      :mc-cmd:`~mc admin tier add TIER_TYPE` is ``azure``. 
+      This option has no effect for any other ``TIER_TYPE``.
+
+   .. mc-cmd:: credentials-file
+      :option:
+
+      *Required*
+      
+      The credential file for a user on the remote GCS tier. The user must have
+      permission to perform read/write/list operations on the remote bucket
```
Comment (krisis) [inline source/reference/minio-cli/minio-mc-admin/mc-admin-tier.rst:360]:
```suggestion
      permission to perform read/write/list/delete operations on the remote bucket
```

File: source/reference/minio-cli/minio-mc/mc-ilm.rst [hunk 15]
```diff
@@ -208,46 +211,193 @@ Syntax
...
-      Specifying a date that is *prior* to the current date marks all
-      objects covered by the rule for transition.
+      The ISO-8601-formatted calendar date after which MinIO marks an object as
+      eligible for transition to the remote tier. MinIO transitions the object
+      to the configured remote storage tier specified to the 
+      :mc-cmd-option:`~mc ilm add storage-class` once the system host datetime
+      passes that calendar date.
+
+      Specifying a calendar date that is *prior* to the current system host
+      datetime marks all objects covered by the rule for transition.
```
[the remark this answers is not in the document]
Comment (ravindk89) [reply source/reference/minio-cli/minio-mc/mc-ilm.rst:298]: Hm - that would make sense for transition-days, but not transition-date. I'm not sure what I was thinking here, so I'll remove this.

File: source/reference/minio-cli/minio-mc-admin/mc-admin-tier.rst [hunk 16]
```diff
@@ -0,0 +1,387 @@
...
+
+      *Optional*
+      
+      The secret key for a user on the remote S3 tier.
+
+      This option only applies to remote storage tiers with 
+      :mc-cmd:`~mc admin tier add TIER_TYPE` is ``s3``. 
+      This option has no effect for any other ``TIER_TYPE``.
+
+   .. mc-cmd:: account-
```
Comment (krisis) [inline source/reference/minio-cli/minio-mc-admin/mc-admin-tier.rst:330]:
```suggestion
   .. mc-cmd:: account-name
```

File: source/reference/minio-cli/minio-mc/mc-ilm.rst (hunk 15 above)
[the remark this answers is not in the document]
Comment (ravindk89) [reply source/reference/minio-cli/minio-mc/mc-ilm.rst:298]: Hm - that would make sense for transition-days, but not transition-date. I'm not sure what I was thinking here, so I'll remove this.

File: source/lifecycle-management/transition-objects-to-gcs.rst [hunk 17]
```diff
@@ -0,0 +1,193 @@
...
+
+Required GCS Permissions
+~~~~~~~~~~~~~~~~~~~~~~~~
+
+Object transition lifecycle management rules require additional permissions on
+the remote storage tier. Specifically, MinIO requires the 
+:abbr:`GCS (Google Cloud Storage)` credentials provide read, write, list, and 
+delete permissions for the remote bucket.
+
+Defer to the `GCS IAM permissions
```
Comment (krisis) [inline source/lifecycle-management/transition-objects-to-gcs.rst:73]:
```suggestion
Refer to the `GCS IAM permissions
```

File: source/lifecycle-management/transition-objects-to-s3.rst [hunk 18]
```diff
@@ -0,0 +1,211 @@
...
+For example, the following policy provides the necessary permission
+for transitioning objects into and out of the remote tier:
+
+.. literalinclude:: /extra/examples/LifecycleManagementUser.json
+   :language: json
+   :class: copyable
+
+Modify the ``Resource`` for the bucket into which MinIO tiers objects.
+
+Defer to the :aws-docs:`Amazon S3 Permissions
```
Comment (krisis) [inline source/lifecycle-management/transition-objects-to-s3.rst:82]:
```suggestion
Refer to the :aws-docs:`Amazon S3 Permissions 
```

**Review by krisis — APPROVED** [review]

**Review by poornas — APPROVED** [review]
LGTM

**Comment by ravindk89** [comment]
Holding off on merging until we have the latest release of `minio` + and the `mc` changes from the pending PR on `mc admin tier`.

**Pull request merged**